File: android/src/main/java/com/aoneahsan/capacitor_auth_manager/MicrosoftAuthProvider.java
package com.aoneahsan.capacitor_auth_manager;

import android.app.Activity;
import android.content.Context;

import com.getcapacitor.JSObject;
import com.microsoft.identity.client.AcquireTokenParameters;
import com.microsoft.identity.client.AcquireTokenSilentParameters;
import com.microsoft.identity.client.AuthenticationCallback;
import com.microsoft.identity.client.IAccount;
import com.microsoft.identity.client.IAuthenticationResult;
import com.microsoft.identity.client.IPublicClientApplication;
import com.microsoft.identity.client.ISingleAccountPublicClientApplication;
import com.microsoft.identity.client.Prompt;
import com.microsoft.identity.client.PublicClientApplication;
import com.microsoft.identity.client.SilentAuthenticationCallback;
import com.microsoft.identity.client.exception.MsalClientException;
import com.microsoft.identity.client.exception.MsalException;
import com.microsoft.identity.client.exception.MsalServiceException;
import com.microsoft.identity.client.exception.MsalUiRequiredException;

import org.json.JSONArray;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

public class MicrosoftAuthProvider implements BaseAuthProvider {
    private static final String TAG = "MicrosoftAuthProvider";

    private final Context context;
    private final Activity activity;
    private final AuthStorage storage;
    private final AuthLogger logger;

    private String clientId;
    private String authority;
    private String redirectUri;
    private List<String> scopes;
    private ISingleAccountPublicClientApplication msalApplication;
    private JSObject currentUser;
    private IAccount currentAccount;

    public MicrosoftAuthProvider(Context context, Activity activity, AuthStorage storage, AuthLogger logger) {
        this.context = context;
        this.activity = activity;
        this.storage = storage;
        this.logger = logger;
        this.scopes = new ArrayList<>(Arrays.asList("openid", "profile", "email"));
    }

    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    public void setAuthority(String authority) {
        this.authority = authority;
    }

    public void setRedirectUri(String redirectUri) {
        this.redirectUri = redirectUri;
    }

    public void setScopes(List<String> scopes) {
        if (scopes != null && !scopes.isEmpty()) {
            this.scopes = scopes;
        }
    }

    @Override
    public void initialize(CapacitorAuthManager.AuthCallback<Void> callback) {
        logger.info("Initializing Microsoft auth provider");

        if (clientId == null || clientId.isEmpty()) {
            callback.onResult(CapacitorAuthManager.AuthResult.error(
                    new Exception("Microsoft client ID is required")));
            return;
        }

        try {
            // Build MSAL configuration JSON
            String configJson = buildMsalConfig();

            PublicClientApplication.createSingleAccountPublicClientApplication(
                    context,
                    configJson,
                    new IPublicClientApplication.ISingleAccountApplicationCreatedListener() {
                        @Override
                        public void onCreated(ISingleAccountPublicClientApplication application) {
                            msalApplication = application;
                            loadAccount(callback);
                        }

                        @Override
                        public void onError(MsalException exception) {
                            logger.error("Failed to create MSAL application", exception);
                            callback.onResult(CapacitorAuthManager.AuthResult.error(exception));
                        }
                    }
            );
        } catch (Exception e) {
            logger.error("Failed to initialize Microsoft auth", e);
            callback.onResult(CapacitorAuthManager.AuthResult.error(e));
        }
    }

    private String buildMsalConfig() {
        StringBuilder config = new StringBuilder();
        config.append("{");
        config.append("\"client_id\": \"").append(clientId).append("\",");

        if (authority != null && !authority.isEmpty()) {
            config.append("\"authorities\": [{");
            config.append("\"type\": \"AAD\",");
            config.append("\"authority_url\": \"").append(authority).append("\",");
            config.append("\"default\": true");
            config.append("}],");
        } else {
            config.append("\"authorities\": [{");
            config.append("\"type\": \"AAD\",");
            config.append("\"authority_url\": \"https://login.microsoftonline.com/common\",");
            config.append("\"default\": true");
            config.append("}],");
        }

        if (redirectUri != null && !redirectUri.isEmpty()) {
            config.append("\"redirect_uri\": \"").append(redirectUri).append("\",");
        }

        config.append("\"account_mode\": \"SINGLE\"");
        config.append("}");

        return config.toString();
    }

    private void loadAccount(CapacitorAuthManager.AuthCallback<Void> callback) {
        if (msalApplication == null) {
            callback.onResult(CapacitorAuthManager.AuthResult.success(null));
            return;
        }

        msalApplication.getCurrentAccountAsync(new ISingleAccountPublicClientApplication.CurrentAccountCallback() {
            @Override
            public void onAccountLoaded(IAccount activeAccount) {
                if (activeAccount != null) {
                    currentAccount = activeAccount;
                    currentUser = createUserObject(activeAccount);
                    logger.info("Restored previous Microsoft session");
                }
                callback.onResult(CapacitorAuthManager.AuthResult.success(null));
            }

            @Override
            public void onAccountChanged(IAccount priorAccount, IAccount currentAccount) {
                if (currentAccount != null) {
                    MicrosoftAuthProvider.this.currentAccount = currentAccount;
                    MicrosoftAuthProvider.this.currentUser = createUserObject(currentAccount);
                }
                callback.onResult(CapacitorAuthManager.AuthResult.success(null));
            }

            @Override
            public void onError(MsalException exception) {
                logger.error("Failed to load Microsoft account", exception);
                callback.onResult(CapacitorAuthManager.AuthResult.success(null));
            }
        });
    }

    @Override
    public void signIn(JSObject credentials, JSObject options, CapacitorAuthManager.AuthCallback<JSObject> callback) {
        logger.info("Starting Microsoft sign-in");

        if (msalApplication == null) {
            callback.onResult(CapacitorAuthManager.AuthResult.error(
                    new Exception("Microsoft Auth not initialized. Call initialize first.")));
            return;
        }

        String[] scopeArray = scopes.toArray(new String[0]);

        AcquireTokenParameters parameters = new AcquireTokenParameters.Builder()
                .startAuthorizationFromActivity(activity)
                .withScopes(Arrays.asList(scopeArray))
                .withPrompt(Prompt.SELECT_ACCOUNT)
                .withCallback(new AuthenticationCallback() {
                    @Override
                    public void onCancel() {
                        logger.warn("Microsoft sign-in cancelled by user");
                        callback.onResult(CapacitorAuthManager.AuthResult.error(
                                new Exception("User cancelled sign-in")));
                    }

                    @Override
                    public void onSuccess(IAuthenticationResult authenticationResult) {
                        handleAuthResult(authenticationResult, callback);
                    }

                    @Override
                    public void onError(MsalException exception) {
                        logger.error("Microsoft sign-in failed", exception);
                        callback.onResult(CapacitorAuthManager.AuthResult.error(exception));
                    }
                })
                .build();

        msalApplication.acquireToken(parameters);
    }

    private void handleAuthResult(IAuthenticationResult authResult, CapacitorAuthManager.AuthCallback<JSObject> callback) {
        currentAccount = authResult.getAccount();
        currentUser = createUserObject(currentAccount);

        JSObject credential = createCredentialObject(authResult);
        storage.saveCredential("microsoft", credential);

        JSObject result = createAuthResultObject(authResult);
        logger.info("Microsoft sign-in successful");
        callback.onResult(CapacitorAuthManager.AuthResult.success(result));
    }

    @Override
    public void signOut(JSObject options, CapacitorAuthManager.AuthCallback<Void> callback) {
        logger.info("Signing out from Microsoft");

        if (msalApplication == null) {
            callback.onResult(CapacitorAuthManager.AuthResult.success(null));
            return;
        }

        msalApplication.signOut(new ISingleAccountPublicClientApplication.SignOutCallback() {
            @Override
            public void onSignOut() {
                currentUser = null;
                currentAccount = null;
                storage.deleteCredential("microsoft");
                callback.onResult(CapacitorAuthManager.AuthResult.success(null));
            }

            @Override
            public void onError(MsalException exception) {
                logger.error("Microsoft sign-out failed", exception);
                callback.onResult(CapacitorAuthManager.AuthResult.error(exception));
            }
        });
    }

    @Override
    public void getCurrentUser(CapacitorAuthManager.AuthCallback<JSObject> callback) {
        if (currentUser != null) {
            callback.onResult(CapacitorAuthManager.AuthResult.success(currentUser));
        } else if (msalApplication != null) {
            msalApplication.getCurrentAccountAsync(new ISingleAccountPublicClientApplication.CurrentAccountCallback() {
                @Override
                public void onAccountLoaded(IAccount activeAccount) {
                    if (activeAccount != null) {
                        currentAccount = activeAccount;
                        currentUser = createUserObject(activeAccount);
                        callback.onResult(CapacitorAuthManager.AuthResult.success(currentUser));
                    } else {
                        callback.onResult(CapacitorAuthManager.AuthResult.success(null));
                    }
                }

                @Override
                public void onAccountChanged(IAccount priorAccount, IAccount currentAccount) {
                    if (currentAccount != null) {
                        MicrosoftAuthProvider.this.currentAccount = currentAccount;
                        MicrosoftAuthProvider.this.currentUser = createUserObject(currentAccount);
                        callback.onResult(CapacitorAuthManager.AuthResult.success(MicrosoftAuthProvider.this.currentUser));
                    } else {
                        callback.onResult(CapacitorAuthManager.AuthResult.success(null));
                    }
                }

                @Override
                public void onError(MsalException exception) {
                    callback.onResult(CapacitorAuthManager.AuthResult.success(null));
                }
            });
        } else {
            callback.onResult(CapacitorAuthManager.AuthResult.success(null));
        }
    }

    @Override
    public void refreshToken(JSObject options, CapacitorAuthManager.AuthCallback<JSObject> callback) {
        logger.info("Refreshing Microsoft token");

        if (msalApplication == null || currentAccount == null) {
            callback.onResult(CapacitorAuthManager.AuthResult.error(
                    new Exception("Microsoft Auth not initialized or no account")));
            return;
        }

        String[] scopeArray = scopes.toArray(new String[0]);

        AcquireTokenSilentParameters parameters = new AcquireTokenSilentParameters.Builder()
                .fromAuthority(currentAccount.getAuthority())
                .withScopes(Arrays.asList(scopeArray))
                .forAccount(currentAccount)
                .forceRefresh(true)
                .withCallback(new SilentAuthenticationCallback() {
                    @Override
                    public void onSuccess(IAuthenticationResult authenticationResult) {
                        JSObject credential = createCredentialObject(authenticationResult);
                        storage.saveCredential("microsoft", credential);

                        JSObject result = createAuthResultObject(authenticationResult);
                        result.put("operationType", "refresh");
                        callback.onResult(CapacitorAuthManager.AuthResult.success(result));
                    }

                    @Override
                    public void onError(MsalException exception) {
                        if (exception instanceof MsalUiRequiredException) {
                            // Need interactive sign-in
                            signIn(null, null, callback);
                        } else {
                            logger.error("Failed to refresh Microsoft token", exception);
                            callback.onResult(CapacitorAuthManager.AuthResult.error(exception));
                        }
                    }
                })
                .build();

        msalApplication.acquireTokenSilentAsync(parameters);
    }

    @Override
    public void isSupported(CapacitorAuthManager.AuthCallback<Boolean> callback) {
        callback.onResult(CapacitorAuthManager.AuthResult.success(true));
    }

    @Override
    public void linkAccount(JSObject credentials, JSObject options, CapacitorAuthManager.AuthCallback<JSObject> callback) {
        signIn(credentials, options, callback);
    }

    @Override
    public void unlinkAccount(CapacitorAuthManager.AuthCallback<Void> callback) {
        signOut(null, callback);
    }

    @Override
    public void getIdToken(boolean forceRefresh, CapacitorAuthManager.AuthCallback<JSObject> callback) {
        if (msalApplication == null || currentAccount == null) {
            callback.onResult(CapacitorAuthManager.AuthResult.error(
                    new Exception("Microsoft Auth not initialized or no account")));
            return;
        }

        String[] scopeArray = scopes.toArray(new String[0]);

        AcquireTokenSilentParameters parameters = new AcquireTokenSilentParameters.Builder()
                .fromAuthority(currentAccount.getAuthority())
                .withScopes(Arrays.asList(scopeArray))
                .forAccount(currentAccount)
                .forceRefresh(forceRefresh)
                .withCallback(new SilentAuthenticationCallback() {
                    @Override
                    public void onSuccess(IAuthenticationResult authenticationResult) {
                        JSObject result = new JSObject();
                        result.put("token", authenticationResult.getAccessToken());
                        callback.onResult(CapacitorAuthManager.AuthResult.success(result));
                    }

                    @Override
                    public void onError(MsalException exception) {
                        callback.onResult(CapacitorAuthManager.AuthResult.error(exception));
                    }
                })
                .build();

        msalApplication.acquireTokenSilentAsync(parameters);
    }

    @Override
    public void revokeAccess(String token, CapacitorAuthManager.AuthCallback<Void> callback) {
        // Microsoft requires web-based sign out for full revocation
        signOut(null, callback);
    }

    private JSObject createUserObject(IAccount account) {
        JSObject user = new JSObject();
        user.put("uid", account.getId());
        user.put("email", account.getUsername());
        user.put("emailVerified", account.getUsername() != null && account.getUsername().contains("@"));
        user.put("displayName", account.getUsername());
        user.put("photoURL", null);
        user.put("phoneNumber", null);
        user.put("isAnonymous", false);
        user.put("tenantId", account.getTenantId());

        // Provider data
        JSONArray providerData = new JSONArray();
        JSObject provider = new JSObject();
        provider.put("providerId", "microsoft.com");
        provider.put("uid", account.getId());
        provider.put("displayName", account.getUsername());
        provider.put("email", account.getUsername());
        provider.put("phoneNumber", null);
        provider.put("photoURL", null);
        providerData.put(provider);
        user.put("providerData", providerData);

        // Metadata
        JSObject metadata = new JSObject();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US);
        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
        metadata.put("lastSignInTime", sdf.format(new Date()));
        user.put("metadata", metadata);

        return user;
    }

    private JSObject createCredentialObject(IAuthenticationResult authResult) {
        JSObject credential = new JSObject();
        credential.put("providerId", "microsoft.com");
        credential.put("signInMethod", "oauth");
        credential.put("accessToken", authResult.getAccessToken());
        credential.put("idToken", authResult.getAccount().getIdToken());
        credential.put("expiresAt", authResult.getExpiresOn().getTime());
        credential.put("tokenType", "Bearer");

        StringBuilder scopeStr = new StringBuilder();
        for (String scope : authResult.getScope()) {
            if (scopeStr.length() > 0) scopeStr.append(" ");
            scopeStr.append(scope);
        }
        credential.put("scope", scopeStr.toString());

        return credential;
    }

    private JSObject createAuthResultObject(IAuthenticationResult authResult) {
        JSObject result = new JSObject();
        result.put("user", createUserObject(authResult.getAccount()));
        result.put("credential", createCredentialObject(authResult));

        JSObject additionalUserInfo = new JSObject();
        additionalUserInfo.put("isNewUser", false);
        additionalUserInfo.put("providerId", "microsoft.com");
        additionalUserInfo.put("username", authResult.getAccount().getUsername());
        result.put("additionalUserInfo", additionalUserInfo);

        result.put("operationType", "signIn");

        return result;
    }
}

File: ios/Plugin/MicrosoftAuthProvider.swift
import Foundation
import Capacitor
import MSAL

class MicrosoftAuthProvider: BaseAuthProvider {
    let provider: AuthProvider = .microsoft
    let storage: AuthStorage
    let logger: AuthLogger

    private var clientId: String?
    private var authority: String?
    private var redirectUri: String?
    private var scopes: [String] = ["openid", "profile", "email"]
    private var msalApplication: MSALPublicClientApplication?
    private var currentUser: AuthUser?

    init(config: AuthProviderConfig, storage: AuthStorage, logger: AuthLogger) {
        self.storage = storage
        self.logger = logger

        // Extract configuration
        if let clientId = config.options["clientId"] as? String {
            self.clientId = clientId
        }
        if let authority = config.options["authority"] as? String {
            self.authority = authority
        }
        if let redirectUri = config.options["redirectUri"] as? String {
            self.redirectUri = redirectUri
        }
        if let scopes = config.options["scopes"] as? [String] {
            self.scopes = scopes
        }
    }

    func initialize(completion: @escaping (Error?) -> Void) {
        logger.info("Initializing Microsoft auth provider")

        guard let clientId = clientId else {
            completion(createAuthError(code: .missingConfiguration, message: "Microsoft client ID is required"))
            return
        }

        do {
            let authorityURL: URL
            if let authority = authority, let url = URL(string: authority) {
                authorityURL = url
            } else {
                authorityURL = URL(string: "https://login.microsoftonline.com/common")!
            }

            let msalAuthority = try MSALAADAuthority(url: authorityURL)

            let config = MSALPublicClientApplicationConfig(
                clientId: clientId,
                redirectUri: redirectUri,
                authority: msalAuthority
            )

            msalApplication = try MSALPublicClientApplication(configuration: config)

            // Check for existing account
            if let accounts = try? msalApplication?.allAccounts(), let account = accounts.first {
                loadUser(from: account)
            }

            completion(nil)
        } catch {
            logger.error("Failed to initialize MSAL: \(error.localizedDescription)")
            completion(createAuthError(code: .internalError, message: error.localizedDescription))
        }
    }

    func signIn(credentials: JSObject?, options: JSObject?, completion: @escaping (AuthResult?, Error?) -> Void) {
        logger.info("Starting Microsoft sign-in")

        guard let application = msalApplication else {
            completion(nil, createAuthError(code: .missingConfiguration, message: "MSAL not initialized"))
            return
        }

        guard let presentingVC = UIApplication.shared.windows.first?.rootViewController else {
            completion(nil, createAuthError(code: .internalError, message: "No presenting view controller available"))
            return
        }

        let webViewParameters = MSALWebviewParameters(authPresentationViewController: presentingVC)
        let interactiveParameters = MSALInteractiveTokenParameters(scopes: scopes, webviewParameters: webViewParameters)

        application.acquireToken(with: interactiveParameters) { [weak self] result, error in
            guard let self = self else { return }

            if let error = error {
                self.logger.error("Microsoft sign-in failed: \(error.localizedDescription)")
                if let nsError = error as NSError?, nsError.domain == MSALErrorDomain {
                    if nsError.code == MSALError.userCanceled.rawValue {
                        completion(nil, self.createAuthError(code: .userCancelled, message: "User cancelled sign-in"))
                        return
                    }
                }
                completion(nil, self.createAuthError(code: .internalError, message: error.localizedDescription))
                return
            }

            guard let result = result else {
                completion(nil, self.createAuthError(code: .internalError, message: "No result returned"))
                return
            }

            let authResult = self.createAuthResult(from: result)
            self.logger.info("Microsoft sign-in successful")
            completion(authResult, nil)
        }
    }

    func signOut(options: JSObject?, completion: @escaping (Error?) -> Void) {
        logger.info("Signing out from Microsoft")

        guard let application = msalApplication else {
            completion(nil)
            return
        }

        do {
            let accounts = try application.allAccounts()
            for account in accounts {
                try application.remove(account)
            }
            currentUser = nil
            storage.deleteCredential(for: provider)
            completion(nil)
        } catch {
            completion(error)
        }
    }

    func getCurrentUser(completion: @escaping (AuthUser?, Error?) -> Void) {
        completion(currentUser, nil)
    }

    func refreshToken(options: JSObject?, completion: @escaping (AuthResult?, Error?) -> Void) {
        logger.info("Refreshing Microsoft token")

        guard let application = msalApplication else {
            completion(nil, createAuthError(code: .missingConfiguration, message: "MSAL not initialized"))
            return
        }

        do {
            let accounts = try application.allAccounts()
            guard let account = accounts.first else {
                completion(nil, createAuthError(code: .invalidUserToken, message: "No account found"))
                return
            }

            let silentParameters = MSALSilentTokenParameters(scopes: scopes, account: account)
            silentParameters.forceRefresh = true

            application.acquireTokenSilent(with: silentParameters) { [weak self] result, error in
                guard let self = self else { return }

                if let error = error {
                    // If silent refresh fails, try interactive
                    self.signIn(credentials: nil, options: nil, completion: completion)
                    return
                }

                guard let result = result else {
                    completion(nil, self.createAuthError(code: .internalError, message: "No result returned"))
                    return
                }

                let authResult = self.createAuthResult(from: result)
                completion(authResult, nil)
            }
        } catch {
            completion(nil, createAuthError(code: .internalError, message: error.localizedDescription))
        }
    }

    func isSupported(completion: @escaping (Bool, Error?) -> Void) {
        completion(true, nil)
    }

    func linkAccount(credentials: JSObject?, options: JSObject?, completion: @escaping (AuthResult?, Error?) -> Void) {
        signIn(credentials: credentials, options: options, completion: completion)
    }

    func unlinkAccount(completion: @escaping (Error?) -> Void) {
        signOut(options: nil, completion: completion)
    }

    func getIdToken(forceRefresh: Bool, completion: @escaping (String?, Error?) -> Void) {
        guard let application = msalApplication else {
            completion(nil, createAuthError(code: .missingConfiguration, message: "MSAL not initialized"))
            return
        }

        do {
            let accounts = try application.allAccounts()
            guard let account = accounts.first else {
                completion(nil, createAuthError(code: .invalidUserToken, message: "No account found"))
                return
            }

            let silentParameters = MSALSilentTokenParameters(scopes: scopes, account: account)
            silentParameters.forceRefresh = forceRefresh

            application.acquireTokenSilent(with: silentParameters) { result, error in
                if let error = error {
                    completion(nil, error)
                } else {
                    completion(result?.idToken, nil)
                }
            }
        } catch {
            completion(nil, error)
        }
    }

    func revokeAccess(token: String?, completion: @escaping (Error?) -> Void) {
        // Microsoft requires web-based sign out to fully revoke
        signOut(options: nil, completion: completion)
    }

    private func loadUser(from account: MSALAccount) {
        currentUser = AuthUser(
            uid: account.homeAccountId?.identifier ?? account.username ?? "",
            email: account.username,
            emailVerified: account.username?.contains("@") ?? false,
            displayName: account.username,
            photoURL: nil,
            phoneNumber: nil,
            isAnonymous: false,
            tenantId: account.homeAccountId?.tenantId,
            providerData: [
                UserInfo(
                    providerId: provider.rawValue,
                    uid: account.homeAccountId?.identifier ?? "",
                    displayName: account.username,
                    email: account.username,
                    phoneNumber: nil,
                    photoURL: nil
                )
            ],
            metadata: UserMetadata(
                creationTime: nil,
                lastSignInTime: ISO8601DateFormatter().string(from: Date()),
                lastRefreshTime: nil
            ),
            refreshToken: nil,
            customClaims: nil
        )
    }

    private func createAuthResult(from msalResult: MSALResult) -> AuthResult {
        let account = msalResult.account

        let authUser = AuthUser(
            uid: account.homeAccountId?.identifier ?? account.username ?? "",
            email: account.username,
            emailVerified: account.username?.contains("@") ?? false,
            displayName: account.username,
            photoURL: nil,
            phoneNumber: nil,
            isAnonymous: false,
            tenantId: account.homeAccountId?.tenantId,
            providerData: [
                UserInfo(
                    providerId: provider.rawValue,
                    uid: account.homeAccountId?.identifier ?? "",
                    displayName: account.username,
                    email: account.username,
                    phoneNumber: nil,
                    photoURL: nil
                )
            ],
            metadata: UserMetadata(
                creationTime: nil,
                lastSignInTime: ISO8601DateFormatter().string(from: Date()),
                lastRefreshTime: nil
            ),
            refreshToken: nil,
            customClaims: nil
        )

        currentUser = authUser

        let credential = AuthCredential(
            providerId: provider.rawValue,
            signInMethod: "oauth",
            accessToken: msalResult.accessToken,
            idToken: msalResult.idToken,
            refreshToken: nil,
            expiresAt: Int(msalResult.expiresOn?.timeIntervalSince1970 ?? 0) * 1000,
            tokenType: "Bearer",
            scope: msalResult.scopes.joined(separator: " "),
            rawNonce: nil
        )

        storage.saveCredential(credential, for: provider)

        return AuthResult(
            user: authUser,
            credential: credential,
            additionalUserInfo: AdditionalUserInfo(
                isNewUser: false,
                providerId: provider.rawValue,
                profile: nil,
                username: account.username
            ),
            operationType: "signIn"
        )
    }
}
